### PR TITLE
Add docker repo override feature to layer0

### LINKED ISSUE #540: Private Docker Repo auth when using Fargate launch type
See discussion below:
https://github.com/quintilesims/layer0/pull/525#discussion_r165734205

### PR DESCRIPTION
**Pull request by sesh-kebab**
**What does this pull request do?**
Adds a new configuration called `docker-repo-override` to l0-setup. This allows the user to specify a mapping of docker image prefixes. For example, the value can be set to: `d.ims.io:12345.dkr.ecr.us-west-2.amazonaws.com`. 

One use case this allows is to allow images to be pulled from an ECR repository without specifying docker credentials.

**How should this be tested?**

1. Create a new l0 instance in a repository other than the carbon account (like the chunnel) with the mapping `d.ims.io:<carbon_aws_account_Id>.dkr.ecr.us-west-2.amazonaws.com`
  (Or you can re-use my instance `seshichunnels`)
2. Create a guestbook deploy
  `l0 deploy create <task definition> guestbook-dpl`
  a. Confirm the deploy correctly replaces the image prefix correctly
  `go run main.go -o json deploy get guestbook-dpl | jq '.[0].dockerrun' -r | base64 -d | jq .`
3. Create a new environment and load balancer 
  `l0 create environment demo-env && l0 loadbalancer create --port 80:80/http demo-env guestbook-lb`
4. Create a service that uses the deploy and load balancer created in the previous steps
  `l0 environment create demo-env`
5. Confirm the service is up and running
  `l0 service create --loadbalancer demo-env:guestbook-lb demo-env guestbook-svc guestbook-dpl:latest`

If the above steps worked, it would've meant that instead of pulling the image via a proxy docker repo frontend (like d.ims.io), the image was pulled directly from ECR by the ecs-agent authenticating using the IAM Role permissions. The image `d.ims.io/quintilesims/guestbook` is also hosted on the carbon account's ECR. So the test confirms that cross-account access will also work.

Notes: An image called `d.ims.io/quintilesims/guestbook:latest` already exists.

You can use the below task definition for step 2:
```
    "AWSEBDockerrunVersion": 2,
    "requiresCompatibilities": [ "EC2" ],
    "containerDefinitions": [
        {
            "name": "guestbook",
            "image": "d.ims.io/quintiles/guestbook",
            "essential": true,
            "memory": 128,
            "portMappings": [
                {
                    "hostPort": 80,
                    "containerPort": 80
                }
            ]
        }
    ]
}
```

**Checklist**
- [x] Unit tests
- ~[ ] Smoke tests (if applicable)~
- ~[ ] System tests (if applicable)~
- [x] Documentation (if applicable)

closes #540 
links https://github.com/quintilesims/d.ims.io/pull/58

### CODE REVIEW

File: common/config/config.go [hunk 1]
```diff
@@ -254,3 +255,24 @@ func ShouldVerifyVersion() bool {
...
+	overrideMap := map[string]string{}
+	for _, s := range strings.Split(v, ",") {
+		if mapping := strings.Split(s, ":"); len(mapping) == 2 {
+			overrideMap[mapping[0]] = mapping[1]
+		}
+	}
+
+	imageParts := strings.Split(image, "/")
+	if len(imageParts) < 2 {
+		return image
```
Comment (tlake) [inline common/config/config.go]: [Nit] We could fail even earlier and avoid a config lookup by moving this block up to the top

**Comment by zpatrick** [comment]
Why wouldn't users just use `12345.dkr.ecr.us-west-2.amazonaws.com` instead of doing the mapping? That just seems like an unnecessary extra step.

**Comment by sesh-kebab** [comment]
The mapping argument is an alternative to using docker config when creating a layer0 instance. So it isn't about a necessary or unecessary step, but giving the user an option in how they want the container instances to pull the images.

File: common/config/config.go (hunk 1 above)
Comment (sesh-kebab) [reply common/config/config.go]: Do you mean the image parts check? I agree. It isn't a case I would expect to fail normally though. More of a mandatory check everytime we use split.

**Comment by zpatrick** [comment]
They can already do that right now, correct? If they use images with the ecr tag in the deploy file?

**Comment by sesh-kebab** [comment]
If they are hosting in carbon, yes.

File: setup/command/init.go [hunk 2]
```diff
@@ -72,6 +76,10 @@ func (f *CommandFactory) Init() cli.Command {
 				overrides[instance.INPUT_AWS_SSH_KEY_PAIR] = v
 			}
 
+			if v := c.String("docker-repo-override"); v != "" {
```
Comment (nehayward) [inline setup/command/init.go]: I wish this just avoided empty strings.

**Comment by sesh-kebab** [comment]
closing due to inactivity and priority. might be worthwhile re-visting at some point.

**Comment by diemonster** [comment]
I think we should consider merging this provided we have some end user documentation

**Comment by diemonster** [comment]
Old PR, closing for now.

**Pull request closed without merging**